storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,24 +3580,347 @@ dberr_t fsp_sys_tablespace_validate()
+  dberr_t err= DB_SUCCESS;
+  buf_block_t *block= fsp_get_header(space, mtr, &err);
+  if (!block)
+    return err;
+  const uint32_t len= flst_get_len(block->page.frame + boffset);
+  fil_addr_t addr= flst_get_first(block->page.frame + boffset);
+  for (uint32_t i= len; i--; )
+  {
+    ut_ad(addr.boffset >= FIL_PAGE_DATA);
+    ut_ad(addr.boffset < block->physical_size() - FIL_PAGE_DATA_END);
+    block= buf_page_get_gen(page_id_t(0, addr.page), 0, RW_SX_LATCH,
+                            nullptr, BUF_GET, mtr, &err);
+    if (!block)
+      return err;
+
+    fil_addr_t next_addr= flst_get_next_addr(block->page.frame + addr.boffset);
+    for (uint32_t i= 0; i < FSP_SEG_INODES_PER_PAGE(srv_page_size); i++)
+    {
+      fseg_inode_t *inode=
+        fsp_seg_inode_page_get_nth_inode(block->page.frame, i);
+      uint16_t offset= uint16_t (inode - block->page.frame);
+      if (mach_read_from_8(FSEG_ID + inode) == 0)
+        continue;
+      if (!inodes->find(addr.page, offset))
+      {
+        err= fseg_inode_free(block, inode, space, mtr);
+        if (err)
+          return err;
+      }
+    }
+    addr= next_addr;
+  }
+  ut_ad(addr.page == FIL_NULL);
+  return err;
+}
+
+/** Remove the unused segment in tablespace. This function
+used only during shrinking of system tablespace
+@param mtr       mini-transaction
+@param shutdown  Called during shutdown
+@return error code */
+dberr_t fil_space_t::garbage_collect(mtr_t *mtr, bool shutdown)
+{
+  if (!shutdown && !trx_sys.is_undo_empty())
+  {
+    sql_print_warning("InnoDB: Can't free the unused segments"
+                      " in system tablespace because a previous shutdown"
+                      " was not with innodb_fast_shutdown=0");
+    return DB_ERROR;
+  }
+
+  ut_a(id == 0);
+  inode_info inodes;
+  dberr_t err= fsp_get_sys_used_segment(&inodes, mtr);
+  if (err)
+    return err;
+
+  mtr->x_lock_space(this);
+  err= fsp_free_unused_seg(this, mtr,
+                           FSP_HEADER_OFFSET + FSP_SEG_INODES_FULL,
+                           &inodes);
+  if (err)
+    return err;
+  err= fsp_free_unused_seg(this, mtr,
+                           FSP_HEADER_OFFSET + FSP_SEG_INODES_FREE,
+                           &inodes);
+  if (err)
+    return err;
+
+  /* Reset rollback segment header page slots because if we upgrade
+  from older version, there is a possiblity that UNDO_CACHED page
+  exist in rollback segment page slots for reuse purpose */
+  for (ulint rseg_id= 0; rseg_id < TRX_SYS_N_RSEGS; rseg_id++)
+  {
+    trx_rseg_t *rseg= &trx_sys.rseg_array[rseg_id];
+    if (rseg->space->id == id)
+    {
+      buf_block_t *block= buf_page_get_gen(rseg->page_id(), 0,
+                                           RW_X_LATCH, nullptr,
+                                           BUF_GET, mtr, &err);
+      if (!block)
+        return err;
+      mtr->memset(block,
+                  TRX_RSEG + TRX_RSEG_UNDO_SLOTS,
+                  TRX_RSEG_N_SLOTS * TRX_RSEG_SLOT_SIZE, 0xff);
```

Comment:
Can you add a comment that says which existing function this is mimicing?

Reply:
This change is wrong. we shouldn't reset the rseg undo slots. Because undo object has already constructed
before this function. If we do that, we need to remove the respective undo objects. I think that is un-necessary.